pkg/yurthub/filter/ingresscontroller/handler.go

```diff
@@ -0,0 +1,179 @@
+		}
+	}
+	endpointsList.Items = items
+
+	return fh.serializer.Encode(endpointsList)
+}
+
+//FilterWatchObject filter the endpoints from watch response object and return the bytes
+func (fh *ingressControllerFilterHandler) StreamResponseFilter(rc io.ReadCloser, ch chan watch.Event) error {
+	defer func() {
+		close(ch)
+	}()
+
+	d, err := fh.serializer.WatchDecoder(rc)
+	if err != nil {
+		klog.Errorf("StreamResponseFilter of ingressControllerFilterHandler ended with error, %v", err)
+		return err
+	}
+	for {
+		watchType, obj, err := d.Decode()
+		if err != nil {
+			return err
+		}
+		var wEvent watch.Event
+		wEvent.Type = watchType
+		endpoints, ok := obj.(*v1.Endpoints)
+		if ok {
+			item := fh.reassembleEndpoint(endpoints)
+			if item == nil {
+				continue
+			}
+			wEvent.Object = item
+		} else {
+			wEvent.Object = obj
+		}
+		klog.V(5).Infof("filter watch decode endpoint: type: %s, obj=%#+v", watchType, endpoints)
+		ch <- wEvent
+	}
+}
+
+// reassembleEndpoints will discard LB service endpoints and filter the valid endpoints to its nodepool
+func (fh *ingressControllerFilterHandler) reassembleEndpoint(endpoints *v1.Endpoints) *v1.Endpoints {
+	svcName := endpoints.Name
+	svc, err := fh.serviceLister.Services(endpoints.Namespace).Get(svcName)
+	if err != nil {
+		klog.Infof("skip reassemble endpoints, failed to get service %s/%s, err: %v", endpoints.Namespace, svcName, err)
+		return endpoints
+	}
+	// discard endpoints if service type is LoadBalancer
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+		klog.Infof("endpoints(%s/%s) of load balancer service is discarded", endpoints.Namespace, svcName)
+		return nil
+	}
+	// filter the endpoints on the node which is in the same nodepool with current node
+	currentNode, err := fh.nodeGetter(fh.nodeName)
+	if err != nil {
+		klog.Infof("skip reassemble endpoints, failed to get current node %s, err: %v", fh.nodeName, err)
+		return endpoints
+	}
+	if nodePoolName, ok := currentNode.Labels[nodepoolv1alpha1.LabelCurrentNodePool]; ok {
+		nodePool, err := fh.nodePoolLister.Get(nodePoolName)
+		if err != nil {
+			klog.Infof("skip reassemble endpoints, failed to get nodepool %s, err: %v", nodePoolName, err)
+			return endpoints
+		}
+		isNodePoolValidEps := false
+		var newEpsA []v1.EndpointAddress
+		for i := range endpoints.Subsets {
+			for j := range endpoints.Subsets[i].Addresses {
+				nodeName := endpoints.Subsets[i].Addresses[j].NodeName
+				if nodeName == nil {
+					//ignore endpoints whose NodeName is not set, for example "kubernetes"
+					continue
+				}
+				if inSameNodePool(*nodeName, nodePool.Status.Nodes) {
+					isNodePoolValidEps = true
+					newEpsA = append(newEpsA, endpoints.Subsets[i].Addresses[j])
+					klog.Infof("endpoints/%s address IP/%s is valid to nodepool %s", svcName, endpoints.Subsets[i].Addresses[j].IP, nodePoolName)
+				}
+			}
+			endpoints.Subsets[i].Addresses = newEpsA
+			newEpsA = nil
+		}
+		if !isNodePoolValidEps {
+			return nil
```

Comment:
after reconstruct endpoints, i think we need to just return `endpoints`, why need to verify `isNodePoolValidEps ` and return nil if `isNodePoolValidEps ` is false?

Reply:
If the endpoints is not valid to the nodepool the node belonging to, it will not be reconstructed and keep unchanged. For these kinds of endpoints I think we should ignore them, so I return nil in my implementation. Not sure whether I have some misunderstanding about it, if yes, please help me to clarify it, thanks!

Reply:
If the endpoints is not valid to the nodepool the node belonging to, it will not be reconstructed and keep unchanged. For these kinds of endpoints I think we should ignore them, so I return nil in my implementation. Not sure whether I have some misunderstanding about it, if yes, please help me to clarify it, thanks!

Reply:
@zzguang Thanks for your explanation, i got it.

Reply:
another question is: endpoints.Subset.NotReadyAddresses should also be reassembled?